File: plugins/command/tunnel/socksproxy/php/proxy.php
<?php
/**
 * 使用UDP服务器处理请求,针对客户端请求应该满足如下格式：
 * +-----+--------+--------+----------+
 * | cmd | sockid | length | data     |
 * +-----+--------+--------+----------+
 * | 1   | 1      | 2      | Variable |
 * +-----+--------+--------+----------+
 * cmd      指定当前动作类型, 可用值：1(connect客户端新建tcp连接), 2(read客户端读取数据), 3(write客户端写入数据), 4(close客户端关闭连接)
 * sockid   指定对应的socket编号，当cmd为2,3时指定要操作的socket, 1时指定新建连接的地址类型（值为4表示ipv4地址，6表示ipv6地址），4时值为0表示关闭UDP服务器（所以socket编号应该从1开始）
 * length   指定数据字段长度，单位字节,大端字节序。cmd为2，4时无意义将置为0
 * data     指定数据部分。当cmd为3时为传输的数据，为1时后俩字节为端口号，之前的为地址（结合sockid判断类型，ipv4 4字节， ipv6 16字节）, 为2，4时无意义将不发送该字段
 * 
 * UDP服务器向客户端请求的响应满足如下格式：
 * +------+--------+----------+
 * | code | length | data     |
 * +------+--------+----------+
 * | 1    | 2      | Variable |
 * +------+--------+----------+
 * code     指定操作是否成功，可用值：0（成功），1（一般性失败），2（连接超时），3（连接拒绝），4（不存在的sockid），7（不支持的cmd）
 * length   指定data字段的长度，大端字节序，单位字节
 * data     根据请求返回数据。
 * 
 * 对于远端的转发请求采用直接转发并存储在临时buf中
 */
//global: $host, $port
set_time_limit(0);
ignore_user_abort(0);
class Connection{
    const BUF_SIZE = 4096;//接收缓冲大小

    public $sock;//stream资源对象
    public $buf = '';
    public $type;
    public $remmote_addr = '';//与sock对应的远端地址，类似127.0.0.1:8080
    public $addr = '';//与UDP服务器交互远程地址类似 127.0.0.1:8080

    public function __construct($sock, $type, $raddr){
        $this->sock = $sock;
        $this->type = $type;
        $this->remote_addr = $raddr;
    }
    public function recvall_to_buf(){//从远端接收数据并存储在buf中
        if(strlen($this->buf)>Connection::BUF_SIZE) return;
        switch($this->type){
            case SOL_UDP:
                if(($buf=stream_socket_recvfrom($this->sock, Connection::BUF_SIZE, 0, $remmote_addraddr)) !== false ){
                    $this->buf .= $buf;
                }
                break;
            case SOL_TCP:
                if(($buf = fread($this->sock, Connection::BUF_SIZE)) !== false){
                    $this->buf .= $buf;
                }
                break;
        }
    }
    public function flush_buf($server){//将buf中的数据写入当前与之对应的客户端连接
        if($this->buf != '' && $this->addr != ''){
            $server->reply(0, $this->buf, $this->addr);
            $this->buf = '';
            $this->addr='';
        }
    }
    public function close($server){
        if($this->addr != ''){
            $server->reply(1, 'CLOSED', $this->addr);
        }
        fclose($this->sock);
    }
    public function sendall($data){//向远端写入数据
        switch($this->type){
            case SOL_UDP:
                return stream_socket_sendto($this->sock, $data, 0, $this->remmote_addr);
            case SOL_TCP:
                return fwrite($this->sock, $data);
        }
        return false;
    }
}
class UDPServer{
    const CONNECT = 1;
    const READ = 2;
    const WRITE = 3;
    const CLOSE = 4;

    public $ret = array('code'=>1, 'msg'=>'');
    private $addr;
    private $server;
    private $connection_list;// {'id':connection}
    public function __construct($addr){
        $this->addr = $addr;
    }
    
    public function init(){
        if(($this->server = stream_socket_server("udp://$this->addr", $errno, $errstr, STREAM_SERVER_BIND))===false){
            $this->ret['code'] = -1;
            if($errno === 0) $this->ret['code'] = -2;
            $this->ret['msg'] = base64_encode("$errno $errstr");
            return false;
        }
        $this->connection_list[0] = new Connection($this->server, SOL_UDP, '');
        return true;
    }

    public function reply($code, $data, $addr){
        $length = strlen($data);
        $buf = pack('Cn', $code, $length);
        stream_socket_sendto($this->server, $buf, 0, $addr);
        stream_socket_sendto($this->server, $data, 0, $addr);
    }

    public function generate_id(){//生成一个可用的socketid,失败返回false
        if(count($this->connection_list) > 255) return false;
        $id = mt_rand(1, 255);
        while(array_key_exists($id, $this->connection_list)){
            $id = mt_rand(1, 255);
        }
        return $id;
    }

    public function test($data){
        $this->ret['code'] = -1;
        $this->ret['msg'] = base64_encode("$data");
        return true;
    }

    public function handle($conn){//处理socket可读状态,返回true则结束UDPserver
        if($this->server === $conn->sock){
            if(($buf = stream_socket_recvfrom($this->server, 4, 0, $addr))===false) return false;
            $cmd = unpack('C', substr($buf, 0, 1))[1];
            $sockid = unpack('C', substr($buf, 1, 1))[1];
            $length = unpack('n', substr($buf, 2, 2))[1];
            switch($cmd){
                case UDPServer::CONNECT:
                    if(($buf = stream_socket_recvfrom($this->server, 6, 0, $addr))!==false){
                        $host = inet_ntop(substr($buf, 0, $length-2));
                        $port = unpack('n', substr($buf, $length-2, 2))[1];
                        if(($sock = stream_socket_client("tcp://[$host]:$port", $errno, $errstr, 3))!==false){
                            $id = $this->generate_id();
                            if($id !== false){
                                $this->connection_list[$id] = new Connection($sock, SOL_TCP, "$host:$port");
                                $this->reply(0, pack('C', $id), $addr);//成功返回sockid
                            }else{
                                $this->reply(1, 'The socket has reached its limit 255', $addr);
                            }
                        }else{
                            $this->reply(2, "$errno $errstr", $addr);
                        }
                    }else{
                        $this->reply(1, 'Recv data failed!', $addr);
                    }
                    break;
                case UDPServer::READ:
                    if(array_key_exists($sockid, $this->connection_list)){
                        $tmp = $this->connection_list[$sockid];
                        $tmp->addr = $addr;//先存地址，后续再将数据传回客户端
                    }else{
                        $this->reply(4, '', $addr);
                    }
                    break;
                case UDPServer::WRITE:
                    if(array_key_exists($sockid, $this->connection_list)){
                        if(($buf = stream_socket_recvfrom($this->server, $length, 0, $addr))!==false){
                            $tmp = $this->connection_list[$sockid];
                            $tmp->sendall($buf);
                            $this->reply(0, '', $addr);
                        }else{
                            $this->reply(1, 'Recv data failed!', $addr);
                        }
                    }else{
                        $this->reply(4, '', $addr);
                    }
                    break;
                case UDPServer::CLOSE:
                    if($sockid === 0){//全部关闭
                        $this->reply(0, '', $addr);
                        $this->close();
                        return true;// 结束
                    }else if(array_key_exists($sockid, $this->connection_list)){
                        $tmp = $this->connection_list[$sockid];
                        unset($this->connection_list[$sockid]);
                        $tmp->close($this);
                        $this->reply(0, '', $addr);
                    }else{
                        $this->reply(4, '', $addr);
                    }
                    break;
                default:
                    $this->reply(7, '', $addr);
            }
        }else{
            $conn->recvall_to_buf();
        }
        return false;
    }

    public function close(){
        foreach($this->connection_list as $id=>$c){//关闭socket
            if($id !== 0)
                $c->close($this);
        }
        fclose($this->server);
    }

    public function start(){
        $w = $e = array();
        while(true){
            $read = array();
            foreach($this->connection_list as $conn){
                $read[] = $conn->sock;
            }
            $r = stream_select($read, $w, $e, NULL);
            if($r){
                foreach($read as $sock){
                    foreach($this->connection_list as $conn){
                        if($conn->sock === $sock){
                            if($this->handle($conn)){
                                return;
                            }
                        }
                    }
                }
            }else{
                $this->ret['code'] = -3;
                $this->close();
                break;
            }
            // 处理待读取数据的客户端
            foreach($this->connection_list as $conn){
                $conn->flush_buf($this);
            }
        }
    }
}

function run($vars){
    extract($vars);
    $server = new UDPServer("$host:$port");
    if($server->init()){
        $server->start();
    }
    return json_encode($server->ret);
}

File: plugins/command/tunnel/socksproxy/proxy.py
from api import utils, logger, Session
import socket
import struct
import json
import base64
import threading
import select

class Connection:
    '''存储连接信息，以及该连接相关操作
    '''

    ACTION_CONNECT = 1
    ACTION_READ = 2
    ACTION_WRITE = 3
    ACTION_CLOSE = 4

    REP_SUCCESS = 0
    REP_FAILED = 1

    def __init__(self, client:socket.socket, rhost: str, rport: int, lhost: str, lport: int, iptype: int, server):
        self.client = client # 已经和本地程序建立连接的socket,已设置为非阻塞模式
        self.rhost = rhost # 远程地址
        self.rport = rport # 远程端口
        self.lhost = lhost # 本地客户端地址与client对应
        self.lport = lport # 本地客户端端口与client对应
        self.iptype = iptype # 为4表示ipv4地址，为6表示ipv6地址
        self.server = server # SocksProxy对象
        self.session = self.server.session
        self.id = 0 # 当转发成功时会从服务端获取，当关闭该连接时置为0，可以通过该值判断连接是否存活

    def is_alive(self)-> bool:
        if self.id:
            return True
        return False

    def __str__(self)-> str:
        return f"({self.lhost}, {self.lport}) => ({self.rhost}, {self.rport})"

    def exec_action(self, action: int, data: bytes=b'')-> int:
        '''向远程UDP服务器执行动作
        '''
        ret = None
        encoding = self.session.options.get_option('encoding').value
        timeout = self.session.options.get_option('timeout').value
        if action == self.ACTION_CONNECT:
            timeout = 20
        elif action == self.ACTION_READ:
            timeout = 0

        ret = self.session.evalfile('action', dict(action=action, shost=self.server.shost, sport=self.server.sport, sockid=self.id, 
            rhost=self.rhost, rport=self.rport, type=self.iptype, data=data), timeout, True)

        if ret is None:
            return self.REP_FAILED

        ret = json.loads(ret)
        if ret['code'] == 1:
            if action == self.ACTION_CONNECT:
                self.id = ret['msg']
                logger.info(f"连接`{self}`创建成功!", True)
            elif action == self.ACTION_CLOSE:
                self.id = 0
                logger.info(f"连接`{self}`关闭!", False)
            elif action == self.ACTION_READ:
                msg = base64.b64decode(ret['msg'].encode())
                self.client.sendall(msg, 0)
            return self.REP_SUCCESS
        elif ret['code'] == -1:
            msg = base64.b64decode(ret['msg'].encode()).decode(encoding, 'ignore')
            logger.error("远程socket发生错误: "+msg)
        elif ret['code'] == -2:
            code = ret['msg']
            logger.error(f"远程错误代码 `{code}`!")
            if action == self.ACTION_READ:
                logger.info("远程数据读取错误!")
        if action == self.ACTION_CONNECT:
            logger.error(f"连接`{self}`创建失败!")
        return self.REP_FAILED


class SocksProxy(threading.Thread):
    '''socks服务端，实现基本功能
    '''

    SOCKS_VERSION_4 = 4
    SOCKS_VERSION_5 = 5
    ATYP_IPV4 = 1
    ATYP_FQDN = 3
    ATYP_IPV6 = 4
    CMD_CONNECT = 1
    CMD_BIND = 2
    CMD_UDP_ASSOCIATE = 3

    def __init__(self, session:Session):
        super().__init__()
        self.session = session
        self.port = 1080
        self.host = '0.0.0.0'
        self.shost = '127.0.0.1' # 远程服务器监听地址
        self.sport = 50000 # 远程服务器监听UDP端口
        self.ldns = False # 是否本地解析域名

        self.connections = []
        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.max_listen_count = 50 # 最大监听数量
        self.upload_buf_size = 4096 # 限制每次上传到远程的数据包大小,为0 时不限制

        self.running = False
        self._lock = threading.Lock()
        self.name = "SocksProxy Local Server"

    def shakehands(self, client:socket.socket, addr:tuple)->Connection:
        '''完成socks握手过程
        '''
        # 客户端握手
        ver, l = struct.unpack('!BB', client.recv(2))
        assert ver in (self.SOCKS_VERSION_4, self.SOCKS_VERSION_5), f"不支持的协议类型`{ver}`"
        assert l > 0, "错误的方法"
        methods = struct.unpack("!"+"B"*l, client.recv(l)) # 备用

        # 无验证方式
        client.sendall(struct.pack('!BB', ver, 0))

        # 处理客户端转发请求
        ver, cmd, rsv, atyp = struct.unpack('!BBBB', client.recv(4))
        assert ver in (self.SOCKS_VERSION_4, self.SOCKS_VERSION_5), f"不支持的协议类型`{ver}`"
        assert rsv == 0, f"Reserved field must be 0, actually `{rsv}`"
        dst_addr = ''
        dst_port = 0
        if atyp == self.ATYP_IPV4:
            dst_addr = socket.inet_ntoa(client.recv(4))
        elif atyp == self.ATYP_IPV6:
            dst_addr = socket.inet_ntop(socket.AF_INET6, client.recv(16))
        elif atyp == self.ATYP_FQDN:
            l = struct.unpack("!B", client.recv(1))[0]
            assert l > 0, f"ATYP is FQDN, but domain length `{l}` is invalid"
            dst_addr = self._handle_domain(client.recv(l).decode())
        else:
            raise AssertionError(f"Unknown atyp `{atyp}`")
        dst_port = struct.unpack('!H', client.recv(2))[0]

        conn = Connection(client, dst_addr, dst_port, addr[0], addr[1], 
            6 if atyp == self.ATYP_IPV6 else 4, self)
        if cmd == self.CMD_CONNECT:
            code = conn.exec_action(Connection.ACTION_CONNECT)
            if code == 0:# 转发成功
                client.sendall(struct.pack("!BBBB", ver, 0, rsv, self.ATYP_IPV4)+
                    socket.inet_pton(socket.AF_INET, self.host)+struct.pack("!H", self.port))
                client.setblocking(False) # 设置非阻塞模式
                return conn
            else:
                client.sendall(struct.pack("!BBBB", ver, code, rsv, self.ATYP_IPV4)+
                    socket.inet_pton(socket.AF_INET, self.host)+struct.pack("!H", self.port))
                raise AssertionError("连接转发失败!")
        elif cmd == self.CMD_BIND:
            pass
        elif cmd == self.CMD_UDP_ASSOCIATE:
            pass
        else:
            raise AssertionError(f"未知的命令字段`{cmd}`")

        return None

    def _handle_domain(self, domain: str)-> str:
        '''处理域名，可能本地解析，也可远程解析
        '''
        if self.ldns:
            return socket.gethostbyname(domain)
        return domain

    def start_remote_udp_server(self):
        '''开启远程UDP服务器，用于转发请求
        '''
        while True:
            ret = self.session.evalfile('proxy', dict(host=self.shost, port=self.sport), 0, True)
            if ret is None:
                break
            
            ret = json.loads(ret)
            if ret['code'] == 1:
                break
            elif ret['code'] in (-1, -2):
                msg = base64.b64decode(ret['msg'].encode()).decode(self.session.options.get_option('encoding').value, 'ignore')
                logger.error("远程转发服务错误: "+msg)
                if ret['code'] == -2:
                    logger.error(f"远程转发服务端口（UDP）`{self.sport}`绑定失败!")
            elif ret['code'] == -3:
                logger.error("远程转发服务错误: 未知的错误原因导致`select`函数执行退出")
            break
        
        with self._lock:
            self.running = False
            try:
                self.server.shutdown(socket.SHUT_RDWR)
            except:
                pass
            self.server.close()

    def close(self):
        with self._lock:
            if self.running:
                encoding = self.session.options.get_option('encoding').value
                ret = self.session.evalfile('action', dict(action=Connection.ACTION_CLOSE, shost=self.shost, sport=self.sport, sockid=0), find_dir=True)
                if ret is None:
                    logger.error("远程转发服务关闭失败!")
                    return
                ret = json.loads(ret)
                if ret['code'] == 1:
                    for conn in self.connections:
                        conn.client.close()
                    self.running = False
                    try:
                        self.server.shutdown(socket.SHUT_RDWR)
                    except:
                        pass
                    self.server.close()
                    logger.info("socks正向转发服务已关闭!")
                    return
                elif ret['code'] == -1:
                    msg = base64.b64decode(ret['msg'].encode()).decode(encoding, 'ignore')
                    logger.error(msg)
                elif ret['code'] == -2:
                    code = ret['msg']
                    logger.error(f"远程错误代码 `{code}`!")

                logger.error("远程转发服务关闭失败!")
                
    def reader(self, conn: Connection):
        '''从远端读
        '''
        while conn.is_alive() and self.running:
            conn.exec_action(Connection.ACTION_READ)

    def writer(self, conn: Connection):
        '''向远端写
        '''
        read = [conn.client, ]
        while conn.is_alive() and self.running:
            r, w, e = select.select(read, [], [])
            if not r:
                logger.error(f"Reader thread exit on `{conn}`")
                conn.exec_action(Connection.ACTION_CLOSE)
                break
            writebuf = b''
            while True:
                tmp = None
                try:
                    tmp = conn.client.recv(4096)
                except BlockingIOError:
                    pass
                except OSError:
                    conn.exec_action(Connection.ACTION_CLOSE)
                    break
                if not tmp:
                    break
                writebuf += tmp
            while writebuf: # 限制每次上传的大小
                block = writebuf[:self.upload_buf_size] if self.upload_buf_size > 0 else writebuf
                writebuf = writebuf[self.upload_buf_size:] if self.upload_buf_size > 0 else False
                conn.exec_action(Connection.ACTION_WRITE, block)

    def run(self):
        '''开启服务
        '''
        self.running = True
        try:
            self.server.bind((self.host, self.port))
        except:
            logger.error(f"本地转发服务绑定`{self.host}, {self.port}`失败!")
            return
        self.server.listen(self.max_listen_count)
        
        thread_list = []
        udp_server_thread = threading.Thread(target=self.start_remote_udp_server, name="SocksProxy Remote UDP Server")
        thread_list.append(udp_server_thread)
        udp_server_thread.setDaemon(True)
        udp_server_thread.start()

        while self.running:
            try:
                client, addr = self.server.accept()
                logger.info(f"来自`{addr}`的连接", True)
            except OSError as e:
                break
            try:
                conn = self.shakehands(client, addr)
            except (AssertionError, OSError) as e:
                logger.error(e)
                client.close()
            else:
                logger.info(f"来自`{addr}`的连接握手成功!", True)
                read_thread = threading.Thread(target=self.reader, args=(conn, ), name=f"SocksProxy read thread on `{conn}`")
                write_thread = threading.Thread(target=self.writer, args=(conn, ), name=f"SocksProxy write thread on `{conn}`")
                thread_list.append(read_thread)
                thread_list.append(write_thread)
                read_thread.setDaemon(True)
                write_thread.setDaemon(True)
                read_thread.start()
                write_thread.start()

        for t in thread_list:
            t.join(5)
            if t.is_alive():
                utils.kill_thread(t.ident)
        logger.info("清理完毕!")
